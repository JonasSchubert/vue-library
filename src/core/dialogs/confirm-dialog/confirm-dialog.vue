<template>
  <v-dialog :max-width="maxWidth" persistent v-model="showDialog">
    <v-card>
      <v-card-title class="headline">{{ translate ? $t(title) : title }}</v-card-title>

      <v-card-text>{{ translate ? !!messageParams && messageParams.length > 0 ? $t(message, messageParams) : $t(message) : message }}</v-card-text>

      <v-card-actions>
        <v-spacer></v-spacer>
        <v-btn @click="cancel" :color="buttons.cancel.textColor" text v-if="!!buttons.cancel">
          <v-icon
            :color="buttons.cancel.iconColor"
            class="mr-1"
            v-if="!!buttons.cancel.icon"
          >{{ buttons.cancel.icon }}</v-icon>
          {{ translate ? $t(buttons.cancel.text) : buttons.cancel.text }}
        </v-btn>
        <v-btn @click="confirm" :color="buttons.confirm.textColor" text>
          <v-icon
            :color="buttons.confirm.iconColor"
            class="mr-1"
            v-if="!!buttons.confirm.icon"
          >{{ buttons.confirm.icon }}</v-icon>
          {{ translate ? $t(buttons.confirm.text) : buttons.confirm.text }}
        </v-btn>
      </v-card-actions>
    </v-card>
  </v-dialog>
</template>

<script lang="ts">
import Vue from "vue";
import { Component } from "vue-property-decorator";
import { ConfirmDialogButtons, ConfirmDialogConfig } from "./index";

@Component({
  name: "confirm-dialog"
})
export default class ConfirmDialog extends Vue {
  public buttons: ConfirmDialogButtons = {
    confirm: {
      icon: undefined,
      iconColor: undefined,
      text: "message.confirm",
      textColor: "success"
    }
  };

  public maxWidth = 360;

  public message = "";

  public messageParams: string[] = [];

  private meta: any = undefined;

  public showDialog = false;

  public title = "";

  public translate = true;

  public cancel(): void {
    this.$emit("cancel", this.meta);
    this.showDialog = false;
  }

  public confirm(): void {
    this.$emit("confirm", this.meta);
    this.showDialog = false;
  }

  public open<T>(config: ConfirmDialogConfig<T>): void {
    this.buttons = config.buttons;
    this.maxWidth = config.maxWidth || 360;
    this.message = config.message;
    this.messageParams = config.messageParams || [];
    this.meta = config.meta;
    this.title = config.title;
    this.showDialog = true;
    this.translate = config.translate !== undefined ? config.translate : true;
  }
}
</script>
