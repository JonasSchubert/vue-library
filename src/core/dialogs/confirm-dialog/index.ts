export * from './confirm-dialog-button';
export * from './confirm-dialog-buttons';
export * from './confirm-dialog-config';
