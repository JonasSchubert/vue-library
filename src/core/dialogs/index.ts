import ConfirmDialog from './confirm-dialog/confirm-dialog.vue';

export const dialogs = [ConfirmDialog];
