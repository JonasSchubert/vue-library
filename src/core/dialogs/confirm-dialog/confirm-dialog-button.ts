import { GaColor } from '@/core/types';

export interface ConfirmDialogButton {
  icon?: string;
  iconColor?: GaColor;
  text: string;
  textColor?: GaColor;
}
