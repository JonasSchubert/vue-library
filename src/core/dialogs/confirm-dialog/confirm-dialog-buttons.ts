import { ConfirmDialogButton } from './confirm-dialog-button';

export interface ConfirmDialogButtons {
  cancel?: ConfirmDialogButton;
  confirm: ConfirmDialogButton;
}
