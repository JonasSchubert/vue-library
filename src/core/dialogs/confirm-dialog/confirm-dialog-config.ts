import { ConfirmDialogButtons } from './confirm-dialog-buttons';

export interface ConfirmDialogConfig<T> {
  buttons: ConfirmDialogButtons;
  maxWidth?: number;
  message: string;
  messageParams?: string[];
  meta?: T;
  title: string;
  translate?: boolean;
}
