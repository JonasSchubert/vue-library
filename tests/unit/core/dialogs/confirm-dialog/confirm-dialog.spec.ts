// Import `shallowMount` from Vue Test Utils and the component being tested
import Vue from 'vue';
import VueI18n from 'vue-i18n';
import { shallowMount } from '@vue/test-utils';
import { ConfirmDialogConfig } from '@/core/dialogs/confirm-dialog';
import ConfirmDialog from '@/core/dialogs/confirm-dialog/confirm-dialog.vue';
import vuetify from '../../../../mocks/vuetify.plugin';

Vue.use(VueI18n);
const i18n: VueI18n = new VueI18n({});

// Mount the component
const wrapper = shallowMount(ConfirmDialog, {
  i18n,
  vuetify
});

describe('ConfirmDialog', () => {
  test('sets the correct default data when mounted', () => {
    // Assert
    const defaultData = wrapper.vm.$data;
    expect(defaultData.buttons).toMatchSnapshot({
      confirm: {
        icon: undefined,
        iconColor: undefined,
        text: 'message.confirm',
        textColor: 'success'
      }
    });
    expect(defaultData.maxWidth).toBe(360);
    expect(defaultData.message).toBe('');
    expect(defaultData.messageParams).toMatchSnapshot([]);
    expect(defaultData.showDialog).toBeFalsy();
    expect(defaultData.title).toBe('');
    expect(defaultData.translate).toBeTruthy();
  });

  describe('methods', () => {
    describe('cancel', () => {
      test('should set showDialog to false', () => {
        // Arrange
        wrapper.vm.$data.showDialog = true;
        wrapper.vm.$emit = jest.fn();

        // Act
        (wrapper.vm as any).cancel();

        // Assert
        expect(wrapper.vm.$data.showDialog).toBeFalsy();
        expect(wrapper.vm.$emit).toHaveBeenCalledTimes(1);
        expect(wrapper.vm.$emit).toHaveBeenCalledWith('cancel', undefined);
      });
    });

    describe('confirm', () => {
      test('should set showDialog to false', () => {
        // Arrange
        wrapper.vm.$data.showDialog = true;
        wrapper.vm.$emit = jest.fn();

        // Act
        (wrapper.vm as any).confirm();

        // Assert
        expect(wrapper.vm.$data.showDialog).toBeFalsy();
        expect(wrapper.vm.$emit).toHaveBeenCalledTimes(1);
        expect(wrapper.vm.$emit).toHaveBeenCalledWith('confirm', undefined);
      });
    });

    describe('open', () => {
      test('should set showDialog to true and set config 1', () => {
        // Arrange
        const confirmDialogConfig: ConfirmDialogConfig<any> = {
          buttons: {
            cancel: {
              icon: 'mdi-exit',
              iconColor: 'accent',
              text: 'message.cancel',
              textColor: 'accent'
            },
            confirm: {
              icon: undefined,
              iconColor: undefined,
              text: 'message.ok',
              textColor: 'success'
            }
          },
          maxWidth: 340,
          message: 'message.ok',
          title: 'message.success'
        };
        wrapper.vm.$data.showDialog = false;

        // Act
        (wrapper.vm as any).open(confirmDialogConfig);

        // Assert
        const defaultData = wrapper.vm.$data;
        expect(defaultData.buttons).toMatchSnapshot({
          cancel: {
            icon: 'mdi-exit',
            iconColor: 'accent',
            text: 'message.cancel',
            textColor: 'accent'
          },
          confirm: {
            icon: undefined,
            iconColor: undefined,
            text: 'message.ok',
            textColor: 'success'
          }
        });
        expect(defaultData.maxWidth).toBe(340);
        expect(defaultData.message).toBe('message.ok');
        expect(defaultData.messageParams).toMatchSnapshot([]);
        expect(defaultData.showDialog).toBeTruthy();
        expect(defaultData.title).toBe('message.success');
        expect(defaultData.translate).toBeTruthy();
      });

      test('should set showDialog to true and set config 2', () => {
        // Arrange
        const confirmDialogConfig: ConfirmDialogConfig<any> = {
          buttons: {
            cancel: {
              icon: 'mdi-exit',
              iconColor: 'accent',
              text: 'message.cancel',
              textColor: 'accent'
            },
            confirm: {
              icon: undefined,
              iconColor: undefined,
              text: 'message.ok',
              textColor: 'success'
            }
          },
          message: 'message.ok',
          title: 'message.success',
          translate: false
        };
        wrapper.vm.$data.showDialog = false;

        // Act
        (wrapper.vm as any).open(confirmDialogConfig);

        // Assert
        const defaultData = wrapper.vm.$data;
        expect(defaultData.buttons).toMatchSnapshot({
          cancel: {
            icon: 'mdi-exit',
            iconColor: 'accent',
            text: 'message.cancel',
            textColor: 'accent'
          },
          confirm: {
            icon: undefined,
            iconColor: undefined,
            text: 'message.ok',
            textColor: 'success'
          }
        });
        expect(defaultData.maxWidth).toBe(360);
        expect(defaultData.message).toBe('message.ok');
        expect(defaultData.messageParams).toMatchSnapshot([]);
        expect(defaultData.showDialog).toBeTruthy();
        expect(defaultData.title).toBe('message.success');
        expect(defaultData.translate).toBeFalsy();
      });
    });
  });
});
