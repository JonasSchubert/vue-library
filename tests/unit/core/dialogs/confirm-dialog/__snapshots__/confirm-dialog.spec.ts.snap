// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmDialog methods open should set showDialog to true and set config 1 1`] = `
Object {
  "cancel": Object {
    "icon": "mdi-exit",
    "iconColor": "accent",
    "text": "message.cancel",
    "textColor": "accent",
  },
  "confirm": Object {
    "icon": undefined,
    "iconColor": undefined,
    "text": "message.ok",
    "textColor": "success",
  },
}
`;

exports[`ConfirmDialog methods open should set showDialog to true and set config 1 2`] = `Object {}`;

exports[`ConfirmDialog methods open should set showDialog to true and set config 2 1`] = `
Object {
  "cancel": Object {
    "icon": "mdi-exit",
    "iconColor": "accent",
    "text": "message.cancel",
    "textColor": "accent",
  },
  "confirm": Object {
    "icon": undefined,
    "iconColor": undefined,
    "text": "message.ok",
    "textColor": "success",
  },
}
`;

exports[`ConfirmDialog methods open should set showDialog to true and set config 2 2`] = `Object {}`;

exports[`ConfirmDialog sets the correct default data when mounted 1`] = `
Object {
  "confirm": Object {
    "icon": undefined,
    "iconColor": undefined,
    "text": "message.confirm",
    "textColor": "success",
  },
}
`;

exports[`ConfirmDialog sets the correct default data when mounted 2`] = `Object {}`;
